keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,191 @@
+(https://github.com/kubernetes/kubernetes/issues/59848) is still open because
+reflectors default to resourceVersion=”0” for their initial list requests. If
+the reflectors instead use a consistent read for their initial list request,
+they could not "going back in time" when components are restarted, which can
+curently happen if the initial list request is served from a stale watch cache
+with data much older than the reflector has previously observed or if the
+api-server or etcd are partitioned.
+
+We have held off on switching reflectors to using consistent read for the
+initial list, even though we know it is more correct, due to concerns with the
+impact on large scale use cases. But if we serve consistent reads from cache,
+there would be very little difference in scalability to how the kube-apiserver
+serves the resourceVersion="0" list requests from reflectors today.
+
+### Goals
+
+- Serve consistent reads from the watch cache for etcd 3.4 or newer
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scailability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
+### Non-Goals
+
+- Serve consistent reads from the watch cache for etcd 3.3 or earlier
+
+## Proposal
+
+Allow the etcd minor version to be specified in the kube-apiserver `--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  For backward compatibility, if `--storage-backend=etcd3` is provided, it will treat this as `etcd3.0`.
+
+Since etcd progress notify is only available in etcd 3.4. We would only serve consistent reads from cache if `--storage-backend` is set to `etcd3.4` or higher.
+For etcd 3.3-, we would continue to serve consistent reads from etcd.
+
+Guard this optmization behind `WatchCacheConsistentReads` feature gate, and only server consistent reads from cache if it is enabled.
+
+When an consistent read request is received and the watch cache is enabled:
+- Use a range(limit=0) request to get the current revision from etcd for the resource type being served. Etcd can serve this type of request efficiently, and the resulting revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- If the cache already has the current revision, serve the request from cache
+- If the current revision is not in the cache:
+  - Request a progress notify from etcd on the etcd watch that keeps the cache up-to-date. This is needed because the watch only receives updates for the range of the keyspace being watched. So updates for KVs outside of that range are not observed by the watch, and so the highest revision observed by the watch can be lower than the current revision even if the watch is up-to-date. A successful progress notify solve this by sending an event that will the highest revision observed to the >= the current revision. The cache is then guaranteed to to sufficiently up-to-date to serve a request at the current revision.
```

Comment:
When I'm looking into documentation, for me it looks that the notification is send to every single watcher that exists:
https://github.com/etcd-io/etcd/blob/master/clientv3/watch.go#L74

If that's the case, it's not exactly what we would want...
@jpbetz @jingyih

Reply:
It's all watch channels, but only for the client that requested it (not all clients). Let's check this carefully, but since we create an etcd client per storage backend, we only have one watch per etcd client and this might be fine?

Reply:
I digged a bit deeper into it and I guess i understand it now:
- each watch can set the `WithProgressNotify` option
- progress notifies are send only to the watches that specify it (but to all of those)
- progress notifies are send periodically by the server (default 10m?)
- additionally you can request for sending it and then its send again to all watchers that requested it

It may be obvious to everyone except from me, but I think it would be useful to clarify it in this proposal too (maybe also in etcd documentation?)

Reply:
yes, the 1st~3nd are available in etcd 3.x, the 4th is available in etcd 3.4

Reply:
Thanks, I've added all this to the document.